### PR TITLE
Differentiate betweet internal and external Neokingdom Token

### PR DESCRIPTION
**Pull request by vrde**
Related to: https://github.com/TelediskoDAO/TIPs/tree/faa6959b7e0bd90d28e6fd52521183cb7a866a98/7

### CODE REVIEW

File: contracts/InternalMarket/InternalMarketBase.sol [hunk 1]
```diff
@@ -3,9 +3,15 @@
...
 
 import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import "../ShareholderRegistry/IShareholderRegistry.sol";
 import "../RedemptionController/IRedemptionController.sol";
 import "../PriceOracle/IStdReference.sol";
 
+import "../NeokingdomTokenExternal/INeokingdomTokenExternal.sol";
+import "../NeokingdomToken/INeokingdomToken.sol";
+
+import "hardhat/console.sol";
```
Comment (vrde) [inline contracts/InternalMarket/InternalMarketBase.sol:13]: remove this pls

Comment (sirnicolaz) [reply contracts/InternalMarket/InternalMarketBase.sol:13]: yes please

File: contracts/InternalMarket/InternalMarketBase.sol [hunk 2]
```diff
@@ -156,30 +174,52 @@ contract InternalMarketBase {
...
-            daoToken.transfer(to, amount),
-            "InternalMarketBase: transfer failed"
-        );
+        if (
+            _shareholderRegistry.isAtLeast(
+                _shareholderRegistry.CONTRIBUTOR_STATUS(),
+                from
+            )
+        ) {
+            _beforeWithdraw(from, amount);
```
Comment (vrde) [inline contracts/InternalMarket/InternalMarketBase.sol:215]: What happens when a contributor is forced into investor role while having their tokens in the internal market?

File: contracts/TokenGateway/TokenGateway.sol [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+    function deposit(uint amount) public {
+        // FIXME: should we check if the address is in the shareholders' registry?
+        _deposit(msg.sender, amount);
+    }
+
+    function withdraw(address to, uint amount) public {
+        _withdraw(msg.sender, to, amount);
+    }
+}
```
Comment (sirnicolaz) [inline contracts/TokenGateway/TokenGateway.sol:41]: This file shall be removed, right?

File: contracts/TokenGateway/TokenGatewayBase.sol [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    function _deposit(address from, uint amount) internal virtual {
+        _tokenExternal.transferFrom(from, address(this), amount);
+        _tokenInternal.mint(from, amount);
+    }
+
+    function _withdraw(address from, address to, uint amount) internal virtual {
+        _tokenInternal.burn(from, amount);
+        _tokenExternal.transfer(to, amount);
+    }
+}
```
Comment (sirnicolaz) [inline contracts/TokenGateway/TokenGatewayBase.sol:38]: This file shall be removed, right?

**Comment by sirnicolaz** [comment]
@vrde have you considered keeping all the wrap/unwrap logic inside the NeokingdomToken, in a way that InternalMarket only deals with NeokingdomToken and NeokingdomToken takes care of the communication with NeokingdomTokenExternal? It would make the integration between the 3 components a bit less tentacular I would guess, and encapsulate the responsibilities in a clearer way

**Comment by vrde** [comment]
superseded by #39

**Pull request closed without merging**